### DISCUSSION TITLE
Link prediction on heterogeneous graph with temporal edges

### DISCUSSION THREAD

**Discussion by DavidBenAttar** [pyg-team/pytorch_geometric#8717] [Q&A]
Hi all!

I want to make a link prediction on a heterogeneous graph where the edges used for supervision should be selected by date (after a certain date in the train) and not randomly, to prevent leakage between the message passing edges timestamp and the (positive) supervision edges timestamp.
How would you propose to do it?

**Comment by akihironitta** [comment]
We have recently added support for temporal edge-level sampling, and its example is here: https://github.com/pyg-team/pytorch_geometric/blob/09733b493493083ab7cb09faf60716308e270093/examples/hetero/temporal_link_pred.py

The feature will be released in the next feature release `2.5.0`, but you can already use it by installing nightly or master:
- https://github.com/pyg-team/pytorch_geometric/#nightly-and-master
- https://github.com/pyg-team/pyg-lib/#from-master

**Reply by DavidBenAttar** [reply]
thank you so much for your quick response.
But my question wasn't for the train test split, but concerning the supervision edge creation.
I want to control which edges will be sampled as supervised edges.
for example:
let's assume my training set is with edges from Jan 1st to Jan 29th.
and I want the message passing edges to be from Jan 1st to Jan 20th, and the supervision edges to be from Jan 20th to Jan 29th.

Is there a way I can control it?

**Reply by akihironitta** [reply]
The example script I linked above is not just about the train/val/test splits. It also guarantees that the timestamp of each seed edge (each supervision edge) is always after the timestamp of its neighboring edges by providing `edge_label_time` to `LinkNeighborLoader` so that it doesn't leak future information per sample:

https://github.com/pyg-team/pytorch_geometric/blob/09733b493493083ab7cb09faf60716308e270093/examples/hetero/temporal_link_pred.py#L45

**Reply by DavidBenAttar** [reply]
Your Doc page says that `edge_label_time` must be the same length as edge_label_index.
So basically all the edges will serve us as supervision edges, thus we'll have an overlap between the supervision edges and the message-passing edges. and that's exactly what  I want to avoid.
Correct me if I'm wrong, but I think we want the message-passing edges to be earlier than **all the supervision edges** (negative and positive). otherwise, we'll still leak future information while training the model.

I don't get how this line will prevent what I've just described:
`edge_label_time=data['user', 'movie'].time[train_idx] - 1,`

It simply takes each element in the tensor and decrements it by one. While I was thinking of one date, all the edges with a date earlier than that would be message-passing and all the rest would be supervision-edges.

**Reply by DavidBenAttar** [reply]
@akihironitta WDYT?

**Reply by rusty1s** [reply]
> let's assume my training set is with edges from Jan 1st to Jan 29th.
and I want the message passing edges to be from Jan 1st to Jan 20th, and the supervision edges to be from Jan 20th to Jan 29th.

This should be supported and is exactly what temporal sampling does. In this case, `edge_label_index` would correspond to the edges from 20th to 29th, while `edge_label_time` would hold a timestamp of 20th for all supervision edges.
